pkg/sloop/queries/respayloadquery.go

```diff
@@ -10,45 +10,65 @@ package queries
 import (
 	"encoding/json"
 	"fmt"
+	"github.com/golang/glog"
 	"github.com/salesforce/sloop/pkg/sloop/kubeextractor"
 	"github.com/salesforce/sloop/pkg/sloop/store/typed"
+	"github.com/salesforce/sloop/pkg/sloop/store/untyped"
 	"github.com/salesforce/sloop/pkg/sloop/store/untyped/badgerwrap"
 	"net/url"
 	"time"
 )
 
 type ResPayLoadData struct {
-	PayLoadMap map[int64]string `json:"payloadMap"`
+	PayloadList []PayloadOuput `json:"payloadList"`
+}
+
+type PayloadOuput struct {
+	PayloadKey  string `json:"payloadKey"`
+	PayLoadTime int64  `json:"payloadTime"`
+	Payload     string `json:"payload,omitempty"`
 }
 
 func GetResPayload(params url.Values, t typed.Tables, startTime time.Time, endTime time.Time, requestId string) ([]byte, error) {
 	var watchRes map[typed.WatchTableKey]*typed.KubeWatchResult
-	err := t.Db().View(func(txn badgerwrap.Txn) error {
-		var err2 error
-		var stats typed.RangeReadStats
+	previousKeyFound := false
+	var previousKey *typed.WatchTableKey
+	var previousVal *typed.KubeWatchResult
+	var previousErr error
 
-		selectedNamespace := params.Get(NamespaceParam)
-		selectedName := params.Get(NameParam)
-		selectedKind := params.Get(KindParam)
+	selectedNamespace := params.Get(NamespaceParam)
+	selectedName := params.Get(NameParam)
+	selectedKind := params.Get(KindParam)
 
-		if kubeextractor.IsClustersScopedResource(selectedKind) {
-			selectedNamespace = DefaultNamespace
-		}
+	if kubeextractor.IsClustersScopedResource(selectedKind) {
+		selectedNamespace = DefaultNamespace
+	}
 
-		key := &typed.WatchTableKey{
-			// partition id will be rest, it is ok to leave it as empty string
-			PartitionId: "",
-			Kind:        selectedKind,
-			Namespace:   selectedNamespace,
-			Name:        selectedName,
-			Timestamp:   time.Time{},
-		}
+	err := t.Db().View(func(txn badgerwrap.Txn) error {
+		var err2 error
+		var stats typed.RangeReadStats
 
+		keyComparator := typed.NewWatchTableKeyComparator(selectedKind, selectedNamespace, selectedName, time.Time{})
 		valPredFn := typed.KubeWatchResult_ValPredicateFns(isResPayloadInTimeRange(startTime, endTime))
-		watchRes, _, err2 = t.WatchTable().RangeRead(txn, key, nil, valPredFn, startTime, endTime)
+		watchRes, _, err2 = t.WatchTable().RangeRead(txn, keyComparator, nil, valPredFn, startTime, endTime)
 		if err2 != nil {
 			return err2
 		}
+
+		// get the previous key for those who has same payload but just before startTime
+		seekKey := keyComparator
+		seekKey.PartitionId = untyped.GetPartitionId(startTime)
+		previousKey, err2 = t.WatchTable().GetPreviousKey(txn, seekKey, keyComparator)
+
+		// when err2 is not nil, we will not return err since it is ok we did not find previous key from startTime,
+		// we can continue using the result from rangeRead to proceed the rest payload
+		if err2 == nil {
+			previousVal, previousErr = t.WatchTable().Get(txn, previousKey.String())
```

Comment:
we should check if `previousErr` is not nil and is equal to KeyNotFound then do nothing, otherwise we should report this error.